src/webview/win/mod.rs

```diff
@@ -33,195 +44,253 @@ impl WV for InnerWebView {
+            )
           }
         }
+        w.Navigate(url_string.as_str())?;
+      }
+    }
 
-        controller.put_is_visible(true);
-        let _ = controller_clone.set(controller);
+    controller.SetIsVisible(true)?;
 
-        if let Some(file_drop_handler) = file_drop_handler {
-          let mut file_drop_controller = FileDropController::new();
-          file_drop_controller.listen(hwnd, file_drop_handler);
-          let _ = file_drop_controller_clone.set(file_drop_controller);
-        }
+    let _ = controller_rc.set(controller).expect("set the controller");
+    let _ = webview_rc.set(w).expect("set the webview");
 
-        Ok(())
-      })
-    })?;
+    if let Some(file_drop_handler) = file_drop_handler {
+      let mut file_drop_controller = FileDropController::new();
+      file_drop_controller.listen(hwnd, file_drop_handler);
+      let _ = file_drop_controller_rc.set(file_drop_controller);
+    }
 
     Ok(Self {
-      controller,
-
-      file_drop_controller,
+      controller: controller_rc,
+      webview: webview_rc,
+      file_drop_controller: file_drop_controller_rc,
     })
   }
 
   fn eval(&self, js: &str) -> Result<()> {
-    if let Some(c) = self.controller.get() {
-      let webview = c.get_webview()?;
-      webview.execute_script(js, |_| (Ok(())))?;
+    if let Some(w) = self.webview.get() {
+      let _ = w.ExecuteScriptAsync(js)?;
     }
     Ok(())
   }
 }
 
 impl InnerWebView {
-  pub fn resize(&self, hwnd: *mut c_void) -> Result<()> {
-    let hwnd = hwnd as HWND;
-
+  pub fn resize(&self, hwnd: HWND) -> Result<()> {
     // Safety: System calls are unsafe
     unsafe {
-      let mut rect = std::mem::zeroed();
-      GetClientRect(hwnd, &mut rect);
+      let mut rect = RECT::default();
+      WindowsAndMessaging::GetClientRect(hwnd, &mut rect);
       if let Some(c) = self.controller.get() {
-        c.put_bounds(rect)?;
+        let (width, height) = (rect.right - rect.left, rect.bottom - rect.top);
+        c.SetBounds(Rect {
+          X: 0f32,
+          Y: 0f32,
+          Width: width as f32,
+          Height: height as f32,
+        })?;
       }
     }
 
     Ok(())
   }
 }
+
+/// The WebView2 threading model runs everything on the UI thread, including callbacks which it triggers
+/// with `PostMessage`, and we're using this here because it's waiting for some async operations in WebView2
+/// to finish before starting the main message loop in `WebView::run`. As long as there are no pending
+/// results in `rx`, it will pump Window messages and check for a result after each message is dispatched.
+///
+/// `GetMessage` is a blocking call, so if we want to send results from another thread, senders from other
+/// threads should "kick" the message loop after sending the result by calling `PostThreadMessage` with an
+/// ignorable/unhandled message such as `WM_APP`.
+fn wait_for_async_operation<T>(op: IAsyncOperation<T>) -> Result<T>
```

Comment:
I based this off of an example that didn't use `winit`, but I think ideally this should use an `EventLoop` for consistency/simplicity. I couldn't figure out how to do that with `EventLoop::run` though, because it returns `!` and sees the `Ok(result)` after that as unreachable.

It would also be a good candidate for `async`/`await`, but then I'd have to make the caller in `async` as well and restructure everything up to the main `EventLoop` to use `futures`.

Reply:
Found `EventLoopExtRunReturn::run_return` which does what I want. It's discouraged, but appropriate for this specific scenario.